clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java

```diff
@@ -59,29 +63,66 @@ public PartitionData(long offset, int maxBytes) {
         }
     }
 
+    private static final class TopicAndPartitionData {
+        public final String topic;
+        public final LinkedHashMap<Integer, PartitionData> partitions;
+
+        public TopicAndPartitionData(String topic) {
+            this.topic = topic;
+            this.partitions = new LinkedHashMap<>();
+        }
+
+        public static final List<TopicAndPartitionData> groupByTopicOrdered(Map<TopicPartition, PartitionData> fetchData) {
+            List<TopicAndPartitionData> topics = new ArrayList<>();
+            for (Map.Entry<TopicPartition, PartitionData> topicEntry : fetchData.entrySet()) {
+                String topic = topicEntry.getKey().topic();
+                int partition = topicEntry.getKey().partition();
+                PartitionData partitionData = topicEntry.getValue();
+                if (topics.isEmpty() || topics.get(topics.size() - 1).topic != topic)
+                    topics.add(new TopicAndPartitionData(topic));
+                topics.get(topics.size() - 1).partitions.put(partition, partitionData);
+            }
+            return topics;
+        }
+    }
+
     /**
      * Create a non-replica fetch request
      */
     public FetchRequest(int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData) {
```

Comment:
Since the order is now significant, maybe we should consider changing these constructors so that they either accept `LinkedHashMap` or `List<TopicAndPartitionData>`?